index.d.ts

```diff
@@ -0,0 +1,592 @@
+   * @default 8126
+   */
+  port?: number | string;
+
+  /**
+   * Set an application’s environment e.g. prod, pre-prod, stage.
+   */
+  env?: string;
+
+  /**
+   * Percentage of spans to sample as a float between 0 and 1.
+   * @default 1
+   */
+  sampleRate?: number;
+
+  /**
+   * Interval in milliseconds at which the tracer will submit traces to the agent.
+   * @default 2000
+   */
+  flushInterval?: number;
+
+  /**
+   * Experimental features can be enabled all at once by using true or individually using key / value pairs.
+   * @default {}
+   */
+  experimental?: ExperimentalOptions | boolean;
+
+  /**
+   * Whether to load all built-in plugins.
+   * @default true
+   */
+  plugins?: boolean;
+
+  /**
+   * Custom logger to be used by the tracer (if debug = true),
+   * should support debug() and error() methods
+   * see https://datadog.github.io/dd-trace-js/#custom-logging
+   */
+  logger?: {
+    debug: (message: string) => void;
+    error: (err: Error) => void;
+  };
+
+  /**
+   * Global tags that should be assigned to every span.
+   */
+  tags?: { [key: string]: any };
+}
+
+interface ExperimentalOptions {}
+
+declare class ScopeManager {
+  /**
+   * Get the current active scope or null if there is none.
+   */
+  public active(): Scope | null;
+
+  /**
+   * Activate a new scope wrapping the provided span.
+   *
+   * @param span The span for which to activate the new scope.
+   * @param finishSpanOnClose Whether to automatically finish the span when the scope is closed.
+   */
+  public activate(span: opentracing.Span, finishSpanOnClose?: boolean): Scope;
+}
+
+declare class Scope {
+  /**
+   * Get the span wrapped by this scope.
+   */
+  public span(): opentracing.Span;
+
+  /**
+   * Close the scope, and finish the span if the scope was created with `finishSpanOnClose` set to true.
+   */
+  public close(): void;
+}
+
+/** @hidden */
+interface Plugins {
+  "amqp10": amqp10.Options;
+  "amqplib": amqplib.Options;
+  "bluebird": bluebird.Options;
+  "elasticsearch": elasticsearch.Options;
+  "express": express.Options;
+  "graphql": graphql.Options;
+  "hapi": hapi.Options;
+  "http": http.Options;
+  "ioredis": ioredis.Options;
+  "koa": koa.Options;
+  "memcached": memcached.Options;
+  "mongodb-core": mongodb_core.Options;
+  "mysql": mysql.Options;
+  "mysql2": mysql2.Options;
+  "pg": pg.Options;
+  "q": q.Options;
+  "redis": redis.Options;
+  "restify": restify.Options;
+  "router": router.Options;
+  "when": when.Options;
+}
+
+/**
+ * This plugin automatically instruments the
+ * [amqp10](https://github.com/noodlefrenzy/node-amqp10) module.
+ *
+ * The following spans are automatically created by this integration:
+ *
+ * * **amqp.send**: Span when messages are sent.
+ * * **amqp.receive**: Span when messages are received.
+ *
+ * Please see the available [options](../interfaces/amqp10.options.html) to
+ * configure this plugin.
+ */
+declare namespace amqp10 {
```

Comment:
It is recommended (in the TS/DefinitelyTyped realm) to use `namespace`. Moreover, the TS support in Babel 7 does not support it at all. When people really want namespaces, it's usually solved by utilizing multiple modules.

Reply:
I was not sure which one should be used, but decided to go with `namespace` as it seems to be the construct recommended for "internal modules" according to the [docs](https://www.typescriptlang.org/docs/handbook/namespaces.html).

Do you have any good reference on the difference between the two and when one should be used instead of the other?

Reply:
It's hard to find a clear doc that states you should favour modules over namespaces and I'm sure there are still valid reasons to use namespaces. I personally follow the rule because it's what the DT repo uses through TSLint's recommended settings (which includes https://palantir.github.io/tslint/rules/no-namespace/).

When it comes to truly modelling a module, using `module` (or separate files to reflect modules) is a direct representation of reality. i.e. I could, if I wanted to, import one of the plugin modules. If I wanted TS to know about types of that module, using namespaces wouldn't make that happen. If you don't really want people to import modules, which seems to be the case here (?), then there's also no real point in modelling modules somewhat. Instead I'd just define an interface for each plugin that models the plugin config object you can pass to `init`, [like we did in the DT types for this lib](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/dd-trace/index.d.ts).

Reply:
Perhaps interesting, yet still implicit, is this from [the docs](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html#using-modules):

> Starting with ECMAScript 2015, modules are native part of the language, and should be supported by all compliant engine implementations. Thus, for new projects modules would be the recommended code organization mechanism.

TS always wants to follow the standard but sometimes experiments with future possible additions to the spec. From reading this snippet and various [3rd](https://stackoverflow.com/a/39827997/95397) [party](https://stackoverflow.com/a/44233460/95397) recounts it seems like this is what happened with namespaces and they are no longer encouraged.

Reply:
cc @DanielRosenwasser do you have any input / know of a definitive document on this topic?

Reply:
Thanks for the explanation! I will have to do some more testing, but basically here is what I'm trying to achieve, which namespaces are currently doing well:

* Scope plugin-related classes to the plugin. We may add new classes/constants per plugin, and it also makes for easier to read names.
* Generate TypeDoc documentation. This has been the most problematic part as it's very difficult to control the output.

I am not too worried if we need to use a bad practice to achieve this goal, especially since we are not actively using TypeScript in the project. What I am worried about however is that you mentioned Babel simply doesn't support namespaces. I'll have to experiment with the alternatives you proposed and see if I can get the same behavior as namespaces.

Reply:
Yeah, the Babel part is my main concern too. However, I’m now realizing that as long as you don’t write the libs source in TS it shouldn’t really matter. Babel doesn’t do any type-checking, for that you’re supposed to use tsc, so in this case the `d.ts` file would simply be ignored and it should all be fine.

As for the example of exporting eg constants from the plugin modules, in that scenario I would actually be more inclined to use modules over namespaces, because it means users will be able to import those exports from the modules directly, instead of only through the main module of the package.

Reply:
I see. So basically for an import such as `import "dd-trace/plugins/express"` for example, it would work if there is a file named `express.d.ts`? I assumed this was only possible for actual `.ts` files only.

As a side note, I usually prefer to export everything from the main module to avoid the API being bound to the file structure, but I'm open to export the files directly if it's deemed useful.

Reply:
Correct, or you could use the `declare module “some/path” { }` declaration in your `index.d.ts` file.

You can ofc still also export them from the main module, but I like letting the decision up to the user as more than once have I had an exotic reason to only import a subset of a package.

Reply:
If the original code uses a namespace pattern for organization, namespaces are fine. New code probably shouldn't be authored using namespaces, but `namespace` is probably fine in `.d.ts` files.

Reply:
I will keep namespace for now and will update later if it causes any issues.